src/SamplingSchemes.jl

```diff
@@ -0,0 +1,314 @@
+    pick_a_random_walker = rand(ats[2:end])
+    to_walk = deepcopy(pick_a_random_walker)
+    accept, rate, at, _, _ = MC_nve_walk!(ns_params.mc_steps, to_walk, lj, ns_params.step_size;
+                                        e_demon_tolerance=mc_routine.e_demon_tolerance,
+                                        demon_energy_threshold=mc_routine.demon_energy_threshold,
+                                        demon_gain_threshold=mc_routine.demon_gain_threshold,
+                                        max_add_steps=mc_routine.max_add_steps)
+    @info "acceptance rate: $rate, emax: $emax, is_accepted: $accept"
+    if accept
+        push!(ats, at)
+        popfirst!(ats)
+        update_iter!(liveset)
+        ns_params.fail_count = 0
+        iter = liveset.walkers[1].iter
+    else
+        @warn "Failed to accept MC move"
+        emax = missing
+        ns_params.fail_count += 1
+    end
+    adjust_step_size(ns_params, rate)
+    return iter, emax, liveset, ns_params
+end
+
+"""
+    adjust_step_size(ns_params::NestedSamplingParameters, rate::Float64)
+
+Adjusts the step size of the nested sampling algorithm based on the acceptance rate. 
+    If the acceptance rate is greater than 0.75, the step size is increased by 1%. 
+    If the acceptance rate is less than 0.25, the step size is decreased by 1%.
+
+# Arguments
+- `ns_params::NestedSamplingParameters`: The parameters of the nested sampling algorithm.
+- `rate::Float64`: The acceptance rate of the algorithm.
+
+# Returns
+- `ns_params::NestedSamplingParameters`: The updated parameters with adjusted step size.
+"""
+function adjust_step_size(ns_params::NestedSamplingParameters, rate::Float64)
+    if rate > 0.75
+        ns_params.step_size *= 1.01
+    elseif rate < 0.25
+        ns_params.step_size *= 0.99
+    end
+    return ns_params
+end
+
+
+
+"""
+    nested_sampling_loop!(liveset::AtomWalkers, ns_params::NestedSamplingParameters, n_steps::Int64, mc_routine::MCRoutine; args...)
+
+Perform a nested sampling loop for a given number of steps.
+
+# Arguments
+- `liveset::AtomWalkers`: The initial set of walkers.
+- `ns_params::NestedSamplingParameters`: The parameters for nested sampling.
+- `n_steps::Int64`: The number of steps to perform.
+- `mc_routine::MCRoutine`: The Monte Carlo routine to use.
+
+# Keyword Arguments
+- `args...`: Additional arguments.
+
+# Returns
+- `df`: A DataFrame containing the iteration number and maximum energy for each step.
+- `liveset`: The updated set of walkers.
+- `ns_params`: The updated nested sampling parameters.
+"""
+function nested_sampling_loop!(liveset::AtomWalkers, ns_params::NestedSamplingParameters, n_steps::Int64, mc_routine::MCRoutine; args...)
+    df = DataFrame(iter=Int[], emax=typeof(0.0u"eV")[])
+    for i in 1:n_steps
+        iter, emax, liveset, ns_params = nested_sampling_step!(liveset, ns_params, mc_routine)
+        if ns_params.fail_count >= 10
+            @warn "Failed to accept MC move 10 times in a row. Break!"
```

Comment:
Does meeting this condition stop the run entirely?

Reply:
For this particular function, yes. There is the dispatch below with `mc_routine::MixedMCRoutine`, where `if ns_params.fail_count >= 10` it switches to the backup routine.